cli/internal/upgrade/terraform.go

```diff
@@ -129,91 +123,71 @@ func (u *TerraformUpgrader) PlanTerraformMigrations(ctx context.Context, opts Te
-	}
-
-	if err := u.fileHandler.RemoveAll(constants.TerraformWorkingDir); err != nil {
-		return fmt.Errorf("removing old terraform directory: %w", err)
+		clusterID.AttestationURL = tfOutput.Azure.AttestationURL
 	}
 
-	if err := u.fileHandler.CopyDir(filepath.Join(constants.UpgradeDir, upgradeID, constants.TerraformUpgradeWorkingDir), constants.TerraformWorkingDir); err != nil {
-		return fmt.Errorf("replacing old terraform directory with new one: %w", err)
+	if err := u.fileHandler.RemoveAll(opts.TFWorkspace); err != nil {
+		return clusterid.File{}, fmt.Errorf("removing old terraform directory: %w", err)
 	}
 
-	if err := u.fileHandler.RemoveAll(filepath.Join(constants.UpgradeDir, upgradeID, constants.TerraformUpgradeWorkingDir)); err != nil {
-		return fmt.Errorf("removing terraform upgrade directory: %w", err)
-	}
-
-	if err := u.mergeClusterIDFile(outputFileContents); err != nil {
-		return fmt.Errorf("merging migration output into %s: %w", constants.ClusterIDsFileName, err)
-	}
-
-	return nil
-}
-
-// mergeClusterIDFile merges the output of the migration into the constellation-id.json file.
-func (u *TerraformUpgrader) mergeClusterIDFile(migrationOutput clusterid.File) error {
-	idFile := &clusterid.File{}
-	if err := u.fileHandler.ReadJSON(constants.ClusterIDsFileName, idFile); err != nil {
-		return fmt.Errorf("reading %s: %w", constants.ClusterIDsFileName, err)
+	if err := u.fileHandler.CopyDir(
+		filepath.Join(opts.UpgradeWorkspace, upgradeID, constants.TerraformUpgradeWorkingDir),
+		opts.TFWorkspace,
+	); err != nil {
+		return clusterid.File{}, fmt.Errorf("replacing old terraform directory with new one: %w", err)
 	}
 
-	if err := u.fileHandler.WriteJSON(constants.ClusterIDsFileName, idFile.Merge(migrationOutput), file.OptOverwrite); err != nil {
-		return fmt.Errorf("writing %s: %w", constants.ClusterIDsFileName, err)
+	if err := u.fileHandler.RemoveAll(filepath.Join(opts.UpgradeWorkspace, upgradeID, constants.TerraformUpgradeWorkingDir)); err != nil {
+		return clusterid.File{}, fmt.Errorf("removing terraform upgrade directory: %w", err)
 	}
 
-	return nil
+	return clusterID, nil
 }
 
 type tfClientCommon interface {
-	ShowPlan(ctx context.Context, logLevel terraform.LogLevel, planFilePath string, output io.Writer) error
-	Plan(ctx context.Context, logLevel terraform.LogLevel, planFile string) (bool, error)
+	ShowPlan(ctx context.Context, logLevel terraform.LogLevel, output io.Writer) error
+	Plan(ctx context.Context, logLevel terraform.LogLevel) (bool, error)
 }
 
 // tfResourceClient is a Terraform client for managing cluster resources.
 type tfResourceClient interface {
-	PrepareUpgradeWorkspace(path, oldWorkingDir, newWorkingDir, backupDir string, vars terraform.Variables) error
+	PrepareUpgradeWorkspace(embeddedPath, oldWorkingDir, newWorkingDir, backupDir string, vars terraform.Variables) error
 	CreateCluster(ctx context.Context, provider cloudprovider.Provider, logLevel terraform.LogLevel) (terraform.ApplyOutput, error)
```

Comment:
Is there a reason to undo the recent terraform changes whith stronger type safety? Maybe a mistake during merge

Reply:
Again, I'm not sure what changes you are talking about in this case.
All that is happening here is removal of constant paths in favor of using dynamic workspace paths.
In some cases I simplified some logic, but that has nothing to do with undoing terraform changes for type safety, does it?